IPython/core/guarded_eval.py

```diff
@@ -853,6 +915,69 @@ def dummy_function(*args, **kwargs):
     return None
 
 
+def _infer_return_value(node: ast.FunctionDef, context: EvaluationContext):
+    """Infer the return value(s) of a function by evaluating all return statements."""
+    func_context = context.replace(transient_locals=context.transient_locals.copy())
+    return_values = _collect_return_values(node.body, func_context)
+
+    if not return_values:
+        return None
+    if len(return_values) == 1:
+        return return_values[0]
+
+    types = {type(v) for v in return_values}
+    if len(types) == 1:
+        t = next(iter(types))
+        if t is dict:
+            keys = set()
+            for v in return_values:
+                keys.update(v.keys())
+            return _Duck(
+                attributes=dict.fromkeys(dir(dict())), items={k: None for k in keys}
+            )
```

Comment:
The value of items should be preserved to, so that further chained completion can be performed.

Reply:
Maybe we need a test for it.